Span `flags` are not decoded

When decoding OTLP data encoded as JSON, the `Span.Flags` field is not decoded.

```terminal
$ cat go.mod
module testing/flag

go 1.23.1

require (
	github.com/stretchr/testify v1.9.0
	go.opentelemetry.io/collector/pdata v1.16.0
)

require (
	github.com/davecgh/go-spew v1.1.1 // indirect
	github.com/gogo/protobuf v1.3.2 // indirect
	github.com/json-iterator/go v1.1.12 // indirect
	github.com/kr/text v0.2.0 // indirect
	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
	github.com/modern-go/reflect2 v1.0.2 // indirect
	github.com/pmezard/go-difflib v1.0.0 // indirect
	go.uber.org/multierr v1.11.0 // indirect
	golang.org/x/net v0.26.0 // indirect
	golang.org/x/sys v0.21.0 // indirect
	golang.org/x/text v0.16.0 // indirect
	google.golang.org/genproto/googleapis/rpc v0.0.0-20240604185151-ef581f913117 // indirect
	google.golang.org/grpc v1.66.2 // indirect
	google.golang.org/protobuf v1.34.2 // indirect
	gopkg.in/yaml.v3 v3.0.1 // indirect
)
```

```go
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"go.opentelemetry.io/collector/pdata/pcommon"
	"go.opentelemetry.io/collector/pdata/ptrace"
)

var (
	pSpanA = func() ptrace.Span {
		s := ptrace.NewSpan()
		s.SetTraceID(pcommon.TraceID([16]byte{0x1}))
		s.SetSpanID(pcommon.SpanID([8]byte{0x2}))
		s.SetFlags(1)
		return s
	}()

	pScope = func() pcommon.InstrumentationScope {
		s := pcommon.NewInstrumentationScope()
		s.SetName("test")
		s.SetVersion("v0.0.1")
		return s
	}()

	pScopeSpans = func() ptrace.ScopeSpans {
		s := ptrace.NewScopeSpans()
		pSpanA.CopyTo(s.Spans().AppendEmpty())
		pScope.CopyTo(s.Scope())
		return s
	}()

	pResSpans = func() ptrace.ResourceSpans {
		rs := ptrace.NewResourceSpans()
		pScopeSpans.CopyTo(rs.ScopeSpans().AppendEmpty())
		return rs
	}()

	pTraces = func() ptrace.Traces {
		traces := ptrace.NewTraces()
		pResSpans.CopyTo(traces.ResourceSpans().AppendEmpty())
		return traces
	}()
)

func TestDecode(t *testing.T) {
	var enc ptrace.JSONMarshaler
	b, err := enc.MarshalTraces(pTraces)
	require.NoError(t, err)

	t.Log(string(b))
	assert.Equal(t, data, b)
}

//	{
//	 "resourceSpans": [
//	   {
//	     "resource": {},
//	     "scopeSpans": [
//	       {
//	         "scope": {
//	           "name": "test",
//	           "version": "v0.0.1"
//	         },
//	         "spans": [
//	           {
//	             "traceId": "01000000000000000000000000000000",
//	             "spanId": "0200000000000000",
//	             "flags": 1,
//	           }
//	         ]
//	       }
//	     ]
//	   }
//	 ]
//	}
var data = []byte(`{"resourceSpans":[{"resource":{},"scopeSpans":[{"scope":{"name":"test","version":"v0.0.1"},"spans":[{"traceId":"01000000000000000000000000000000","spanId":"0200000000000000","parentSpanId":"","flags":1,"status":{}}]}]}]}`)

func TestEncode(t *testing.T) {
	var dec ptrace.JSONUnmarshaler
	got, err := dec.UnmarshalTraces(data)
	require.NoError(t, err)
	assert.Equal(t, pTraces, got)
}
```

```terminal
$ go test -v .
=== RUN   TestDecode
    bug_test.go:53: {"resourceSpans":[{"resource":{},"scopeSpans":[{"scope":{"name":"test","version":"v0.0.1"},"spans":[{"traceId":"01000000000000000000000000000000","spanId":"0200000000000000","parentSpanId":"","flags":1,"status":{}}]}]}]}
--- PASS: TestDecode (0.00s)
=== RUN   TestEncode
    bug_test.go:85:
        	Error Trace:	/home/tyler/go/src/testing/flag/bug_test.go:85
        	Error:      	Not equal:
        	            	expected: ptrace.Traces{orig:(*v1.ExportTraceServiceRequest)(0xc000012b28), state:(*internal.State)(0xc000015758)}
        	            	actual  : ptrace.Traces{orig:(*v1.ExportTraceServiceRequest)(0xc000012de0), state:(*internal.State)(0xc000015f88)}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -29,3 +29,3 @@
        	            	         },
        	            	-        Flags: (uint32) 1,
        	            	+        Flags: (uint32) 0,
        	            	         Name: (string) "",
        	Test:       	TestEncode
--- FAIL: TestEncode (0.00s)
FAIL
FAIL	testing/flag	0.005s
FAIL
```

It looks to be a missing case statement here:

https://github.com/open-telemetry/opentelemetry-collector/blob/9811830023e648c8b56008c28bf5459cf1135d38/pdata/ptrace/json.go#L98-L151

--- a/.chloggen/json-decoded-flag.yaml
+++ b/.chloggen/json-decoded-flag.yaml
@@ -1,0 +1,25 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: 'bug_fix'
+
+# The name of the component, or a single word describing the area of concern, (e.g. otlpreceiver)
+component: 'pdata'
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Unmarshal span and link flags from JSON"
+
+# One or more tracking issues or pull requests related to the change
+issues: [11267]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []

--- a/.chloggen/json-decoded-flag.yaml
+++ b/.chloggen/json-decoded-flag.yaml
@@ -7,7 +7,7 @@ change_type: 'bug_fix'
 component: 'pdata'
 
 # A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
-note: "Unmarshal span and link flags from JSON"
+note: "Unmarshal Span and SpanLink flags from JSON"
 
 # One or more tracking issues or pull requests related to the change
 issues: [11267]

--- a/pdata/ptrace/json.go
+++ b/pdata/ptrace/json.go
@@ -112,6 +112,8 @@ func (dest Span) unmarshalJsoniter(iter *jsoniter.Iterator) {
 			if err := dest.orig.ParentSpanId.UnmarshalJSON([]byte(iter.ReadString())); err != nil {
 				iter.ReportError("readSpan.parentSpanId", fmt.Sprintf("parse parent_span_id:%v", err))
 			}
+		case "flags":
+			dest.orig.Flags = iter.ReadUint32()
 		case "name":
 			dest.orig.Name = iter.ReadString()
 		case "kind":
@@ -184,6 +186,8 @@ func (dest SpanLink) unmarshalJsoniter(iter *jsoniter.Iterator) {
 			})
 		case "droppedAttributesCount", "dropped_attributes_count":
 			dest.orig.DroppedAttributesCount = json.ReadUint32(iter)
+		case "flags":
+			dest.orig.Flags = json.ReadUint32(iter)
 		default:
 			iter.Skip()
 		}

--- a/pdata/ptrace/json_test.go
+++ b/pdata/ptrace/json_test.go
@@ -37,6 +37,7 @@ var tracesOTLP = func() Traces {
 	il.SetSchemaUrl("schemaURL")
 	// Add spans.
 	sp := il.Spans().AppendEmpty()
+	sp.SetFlags(1)
 	sp.SetName("testSpan")
 	sp.SetKind(SpanKindClient)
 	sp.SetDroppedAttributesCount(1)
@@ -83,13 +84,14 @@ var tracesOTLP = func() Traces {
 	link.Attributes().PutInt("int", 1)
 	link.Attributes().PutDouble("double", 1.1)
 	link.Attributes().PutEmptyBytes("bytes").FromRaw([]byte("foo"))
+	link.SetFlags(1)
 	// Add another span.
 	sp2 := il.Spans().AppendEmpty()
 	sp2.SetName("testSpan2")
 	return td
 }()
 
-var tracesJSON = `{"resourceSpans":[{"resource":{"attributes":[{"key":"host.name","value":{"stringValue":"testHost"}},{"key":"service.name","value":{"stringValue":"testService"}}],"droppedAttributesCount":1},"scopeSpans":[{"scope":{"name":"scope name","version":"scope version"},"spans":[{"traceId":"0102030405060708090a0b0c0d0e0f10","spanId":"1112131415161718","traceState":"state","parentSpanId":"1112131415161718","name":"testSpan","kind":3,"startTimeUnixNano":"1684617382541971000","endTimeUnixNano":"1684623646539558000","attributes":[{"key":"string","value":{"stringValue":"value"}},{"key":"bool","value":{"boolValue":true}},{"key":"int","value":{"intValue":"1"}},{"key":"double","value":{"doubleValue":1.1}},{"key":"bytes","value":{"bytesValue":"Zm9v"}},{"key":"array","value":{"arrayValue":{"values":[{"intValue":"1"},{"stringValue":"str"}]}}},{"key":"kvList","value":{"kvlistValue":{"values":[{"key":"int","value":{"intValue":"1"}},{"key":"string","value":{"stringValue":"string"}}]}}}],"droppedAttributesCount":1,"events":[{"timeUnixNano":"1684620382541971000","name":"eventName","attributes":[{"key":"string","value":{"stringValue":"value"}},{"key":"bool","value":{"boolValue":true}},{"key":"int","value":{"intValue":"1"}},{"key":"double","value":{"doubleValue":1.1}},{"key":"bytes","value":{"bytesValue":"Zm9v"}}],"droppedAttributesCount":1}],"droppedEventsCount":1,"links":[{"traceId":"0102030405060708090a0b0c0d0e0f10","spanId":"1112131415161718","traceState":"state","attributes":[{"key":"string","value":{"stringValue":"value"}},{"key":"bool","value":{"boolValue":true}},{"key":"int","value":{"intValue":"1"}},{"key":"double","value":{"doubleValue":1.1}},{"key":"bytes","value":{"bytesValue":"Zm9v"}}],"droppedAttributesCount":1}],"droppedLinksCount":1,"status":{"message":"message","code":1}},{"traceId":"","spanId":"","parentSpanId":"","name":"testSpan2","status":{}}],"schemaUrl":"schemaURL"}],"schemaUrl":"schemaURL"}]}`
+var tracesJSON = `{"resourceSpans":[{"resource":{"attributes":[{"key":"host.name","value":{"stringValue":"testHost"}},{"key":"service.name","value":{"stringValue":"testService"}}],"droppedAttributesCount":1},"scopeSpans":[{"scope":{"name":"scope name","version":"scope version"},"spans":[{"traceId":"0102030405060708090a0b0c0d0e0f10","spanId":"1112131415161718","traceState":"state","parentSpanId":"1112131415161718","flags":1,"name":"testSpan","kind":3,"startTimeUnixNano":"1684617382541971000","endTimeUnixNano":"1684623646539558000","attributes":[{"key":"string","value":{"stringValue":"value"}},{"key":"bool","value":{"boolValue":true}},{"key":"int","value":{"intValue":"1"}},{"key":"double","value":{"doubleValue":1.1}},{"key":"bytes","value":{"bytesValue":"Zm9v"}},{"key":"array","value":{"arrayValue":{"values":[{"intValue":"1"},{"stringValue":"str"}]}}},{"key":"kvList","value":{"kvlistValue":{"values":[{"key":"int","value":{"intValue":"1"}},{"key":"string","value":{"stringValue":"string"}}]}}}],"droppedAttributesCount":1,"events":[{"timeUnixNano":"1684620382541971000","name":"eventName","attributes":[{"key":"string","value":{"stringValue":"value"}},{"key":"bool","value":{"boolValue":true}},{"key":"int","value":{"intValue":"1"}},{"key":"double","value":{"doubleValue":1.1}},{"key":"bytes","value":{"bytesValue":"Zm9v"}}],"droppedAttributesCount":1}],"droppedEventsCount":1,"links":[{"traceId":"0102030405060708090a0b0c0d0e0f10","spanId":"1112131415161718","traceState":"state","attributes":[{"key":"string","value":{"stringValue":"value"}},{"key":"bool","value":{"boolValue":true}},{"key":"int","value":{"intValue":"1"}},{"key":"double","value":{"doubleValue":1.1}},{"key":"bytes","value":{"bytesValue":"Zm9v"}}],"droppedAttributesCount":1,"flags":1}],"droppedLinksCount":1,"status":{"message":"message","code":1}},{"traceId":"","spanId":"","parentSpanId":"","name":"testSpan2","status":{}}],"schemaUrl":"schemaURL"}],"schemaUrl":"schemaURL"}]}`
 
 func TestJSONUnmarshal(t *testing.T) {
 	decoder := &JSONUnmarshaler{}

